core/drivers/scmi-msg/clock_generic.c

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <drivers/clk.h>
+#include <drivers/scmi.h>
+#include <drivers/scmi-msg.h>
+#include <kernel/boot.h>
+#include <string.h>
+#include <sys/queue.h>
+
+#define SCMI_MAX_CLK_NAME_LEN	16
+
+/**
+ * struct scmi_clk - Binds an SCMI channel/clock to a core clk reference
+ * @clk:        Core clock reference
+ * @channel_id: SCMI server channel handler exposing the clock
+ * @scmi_id:    SCMI clock domain ID
+ * @enabled:    SCMI clock state
+ * @link:       Reference in SCMI server clock list
+ */
+struct scmi_clk {
+	struct clk *clk;
+	unsigned int channel_id;
+	unsigned int scmi_id;
+	bool enabled;
+	SLIST_ENTRY(scmi_clk) link;
+};
+
+static bool scmi_clk_init_done;
+static SLIST_HEAD(, scmi_clk) scmi_clk_list =
+					SLIST_HEAD_INITIALIZER(scmi_clk_list);
+
+size_t plat_scmi_clock_count(unsigned int channel_id)
+{
+	unsigned int count = 0;
+	unsigned int max_id = 0;
+	struct scmi_clk *clk = NULL;
+
+	SLIST_FOREACH(clk, &scmi_clk_list, link) {
+		if (clk->channel_id == channel_id) {
+			count++;
+			max_id = MAX(max_id, clk->scmi_id);
+		}
+	}
+
+	if (!count)
+		return 0;
+
+	/* IDs are starting from 0 so we need to return max_id + 1 for count */
+	return max_id + 1;
+}
+
+static struct scmi_clk *clk_scmi_get_by_id(unsigned int channel_id,
+					   unsigned int scmi_id)
+{
+	struct scmi_clk *clk = NULL;
+
+	SLIST_FOREACH(clk, &scmi_clk_list, link) {
+		if (clk->channel_id == channel_id &&
+		    clk->scmi_id == scmi_id)
```

Comment:
nitpicking: fits in a signle line. suggestion: also remove braces:
```c
	SLIST_FOREACH(clk, &scmi_clk_list, link)
		if (clk->channel_id == channel_id && clk->scmi_id == scmi_id)
			return clk;

Reply:
Done